doc/specs/#5000 - Process Model 2.0/#5000 - Process Model 2.0.md

```diff
@@ -0,0 +1,882 @@
+process, and move those content processes easily between windows. However, we
+want to make sure that it's also possible to communicate between these
+processes. What if we want to have only a single WT instance, so that whenever
+Wt is run, it creates a tab in the existing window? What if you want to run a
+commandline in a given WT window?
+
+In addition to the concept of Window and Content Processes, we'll also be
+introducing another type of categorization for window processes. These are
+"Monarch" and "Servant" processes. This will allow for the coordination across
+the various windows by the Monarch.
+
+There will only ever be one monarch process at a given time, and every other WT
+window process is a servant process. However, we want this system to be
+redundant, so that if the monarch ever dies, one of the remaining servant
+processes can take over for it. The new monarch will be chosen at random, so
+we'll call this a probabilistic elective monarchy.
+
+Essentially, the probabilistic elective monarchy will work in the following way:
+
+1. We'll introduce a WinRT class (for the purpose of this doc we'll call it
+  `Monarch`), with a unique GUID.
+2. When any window process starts up, it'll first try to register as the server
+   for the `Monarch` WinRT class.
+   - The OS will allow subsequent processes to successfully register as the
+    server for `Monarch` objects, but when someone tries to `create_instance` a
+    `Monarch`, the OS will always create the `Monarch` from the first process to
+    register.
+3. After registering as a server for `Monarch`s, attempt to create a `Monarch`
+   using `winrt::create_instance`.
+4. Using that `Monarch`, ask it for it's PID.
+   - If that PID is the same as the PID of the current process, then the window
+     process knows that it is the monarch.
+   - If that PID is some other process, then we know that we're not currently
+     the monarch.
+     - If we don't currently have an ID assigned to us, then ask the `Monarch`
+       to assign one to us.
+     - If we do have an ID (from a previous monarch), then let the new monarch
+       know that we exist.
+5. If we're a servant process, then `WaitForSingleObject` on a handle to the
+   monarch process. When the current monarch dies, go back to 3. At this point,
+   we might be appointed the new monarch (by whatever process the OS uses to
+   choose who the new server for `Monarch`s is.)
+   - We're suggesting `WaitForSingleObject` here as opposed to having the
+     monarch raise a WinRT event when it's closed, because it's possible that
+     the monarch <!-- is assassinated --> closes unexpectedly, before it has an
+     opportunity to notify the servants.
+
+By this mechanism, the processes will be able to communicate with the Monarch,
+who'll be responsible for managing any inter-process communication between
+windows we might need.
+
+
+#### Scenario: Single Instance Mode

```

Comment:
Does this mean that there will only be one `wt.exe` process that is holding both the window and the content for every tab in "Single Instance Mode?"

I would think that we would still want to use the multiprocess model of a single window process and multiple content processes. 

It's unclear what the actual proposal is here.

Reply:
I've added a bunch more details to this section, and some diagrams, to hopefully clear it up.